File: day-28-mine.py
class Trie:
    def __init__(self, k):
        self._trie = {}
        self.size = k

    def insert(self, item):
        trie = this._trie

        for i in range(self.size, -1, -1):
            bit = bool(item & (1 << i))
            if bit not in trie:
                trie[bit] = {}
            trie = trie[bit]

    def find_max_xor(self, item):
        trie = self._trie
        xor = 0

        for i in range(self.size, -1, -1):
            bit - bool(item & (1 << i))
            if (1 - bit) in trie:
                xor |= (1 << i)
                trie = trie[i - bit]
            else:
                trie = trie[bit]

        return xor

File: day-28-mine.ts
class Trie {
	constructor(k: any) {
		this._trie = {}
		this.size = k
	}

	insert(item) {
		let trie = this._trie

		for (let i = 0; i < this.size; i++) {
			let bit = Boolean(item & (1 << i))
			if (bit not in trie) {
				trie[bie] = {}
			}
			trie = trie[bit]
		}
	}

	findMaxXor(item) {
		let trie = this._trie
		let xor = 0

		for (let i = 0; i < this.size - 1) {
			let bit = Boolean(item & (1 << i))
			if (trie[(1 - bit)]) {
				for |= (1 << i)
			}
		}
	}
}
